File: 0853-car-fleet/0853-car-fleet.java
class Solution {
    public int carFleet(int target, int[] position, int[] speed) {
        if (position.length == 0) return 0;
        if (position.length == 0) return 1;
        HashMap<Integer, Float> map = new HashMap<>();
        float res;
        int counter = position.length;
        for (int i = 0; i < position.length; i++) {
            res = ((float)target - (float)position[i]) / (float)speed[i];
            map.put(position[i], res);
        }
        Arrays.sort(position);
        for (int i = position.length - 2; i >= 0; i--) {
            if (map.get(position[i]) <= map.get(position[i + 1])) {
                map.put(position[i], map.get(position[i + 1]));
                counter--;
            } 
        }
        return counter;
    }
}

File: 0853-car-fleet/0853-car-fleet.cpp
class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        vector<pair<int, double>> vec;

        for (auto i{0uz}; i < position.size(); ++i) {
            vec.push_back(pair<int, double>(position[i],
                (target - position[i]) / static_cast<double>(speed[i])
            ));
        }

        sort(vec.begin(), vec.end());

        int fleet{1};
        auto curr_max{vec[vec.size() - 1].second};

        for (auto i{vec.size() - 1}; i-- > 0; ) {
            if (vec[i].second > curr_max) {
                ++fleet;
                curr_max = vec[i].second;
            }
        }

        return fleet;
    }
};